llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -479,27 +480,78 @@ SUnit *AIEPostRASchedStrategy::pickNodeAndCycle(
 }
 
 int AIEPostRASchedStrategy::getMaxDeltaCycles(const SchedBoundary &Zone) const {
-  assert(!Zone.isTop());
-  if (Zone.getCurrCycle() >= RegionBottomUpCycles - 1)
+  // Top-down scheduling does not support DeltaCycles
+  if (Zone.isTop() || Zone.getCurrCycle() >= RegionBottomUpCycles - 1)
     return 0;
   return std::min({int(RegionBottomUpCycles - 1 - Zone.getCurrCycle()),
                    int(getAIEHazardRecognizer(Zone)->getMaxLookAhead()),
                    BottomUpDelta.getValue()});
 }
 
+unsigned getNumEmittedInstrs(ScheduleDAGMI *DAG, bool IsTop) {
+  if (IsTop)
+    return DAG->top().isValid() ? std::distance(DAG->begin(), DAG->top()) : 0;
+  return DAG->bottom().isValid() ? std::distance(DAG->bottom(), DAG->end()) : 0;
+}
+
+SUnit *AIEPostRASchedStrategy::getNextUnscheduledFixedInstr(
+    const SchedBoundary &Zone) const {
+  if (Zone.isTop())
+    return nullptr;
+  const Region &Reg = InterBlock.getBlockState(CurMBB).getCurrentRegion();
```

Comment:
Please describe what happens here. I'm wondering why we don't run across the SUnits by index. We know where they are, in what order they should be scheduled and what their scheduled cycle should be.

Reply:
I think that the easiest way for the fixed instructions to be interleaved is to have them scheduled as "normally" as possible. That means they get dependencies to other SUnits, and are part of the normal pending/available queues. This is appealing because "free" SUnits can be interleaved seamlessly. The only real place that I have to tweak is here: to determine what goes in the available queue.

Reply:
No worries, I agree on that. I was just thinking that the next SU could be easily found by stepping an iterator across the fixed unit each time we a get a call back that a node is scheduled. Alternatively, that we make all the nodes whose time is due available in one go and have a very simple priority function. I don't think that single choice call adds much benefit if the pending queue is overloaded with all the fixed candidates.